test/functional.jl

```diff
@@ -0,0 +1,108 @@
+module TestFunctional
+
+# nullable equality is a little strange... we need a "is a null with type T"
+# and a "is a nullable with value egal to k"
+
+# these functions are curried for readability
+isnull_typed(u::Nullable, T::Type) = typeof(u).parameters[1] == T && isnull(u)
+isnull_typed(t::Type) = u -> isnull_typed(u, t)
+
+isnullableof(u::Nullable, x) = !isnull(u) && u.value === x
+isnullableof(x) = u -> isnullableof(u, x)
+
+using Base.Test
+
+# getindex
+@test_throws NullException Nullable()[]
+@test_throws NullException Nullable{Int}()[]
+@test Nullable(0)[] == 0
+
+@test_throws BoundsError Nullable()[1]
+@test_throws BoundsError Nullable(0)[0]
+@test_throws BoundsError Nullable(0)[2]
+@test Nullable(0)[1] == 0
+
+# collect
+@test collect(Nullable()) == Union{}[]
```

Comment:
Be careful, `==` doesn't test the type at all. You need a separate test using `typeof`.